File: leetcode/binary_search/3506_Find_Time_Required_to_Eliminate_Bacterial_Strains/codes/bs.py
from typing import List

class Solution:
    def check(self, timeReq: List[int], maxTime: int, splitTime: int) -> bool:
        currentWBC = 1
        curTime = 0
        n = len(timeReq)

        for i in range(n):
            if currentWBC == 0:
                return False
            while curTime + splitTime + timeReq[i] <= maxTime and currentWBC < n - i:
                currentWBC <<= 1
                curTime += splitTime
            currentWBC -= 1
        return True

    def minEliminationTime(self, timeReq: List[int], splitTime: int) -> int:
        timeReq.sort(reverse=True)
        left, right = timeReq[0], timeReq[0] + 20 * splitTime

        while left < right:
            mid = (left + right) // 2
            if self.check(timeReq, mid, splitTime):
                right = mid
            else:
                left = mid + 1
        return left

File: leetcode/binary_search/3506_Find_Time_Required_to_Eliminate_Bacterial_Strains/codes/bs.c
#include <stdio.h>
#include <stdlib.h>

int cmp(const void* a, const void* b) {
    return (*(int*)b - *(int*)a); 
}

int check(int* timeReq, int n, long long maxTime, int splitTime) {
    int currentWBC = 1;
    long long curTime = 0;
    for (int i = 0; i < n; ++i) {
        if (currentWBC == 0) return 0;
        while (curTime + splitTime + timeReq[i] <= maxTime && currentWBC < n - i) {
            currentWBC <<= 1;
            curTime += splitTime;
        }
        currentWBC--;
    }
    return 1;
}

long long minEliminationTime(int* timeReq, int n, int splitTime) {
    qsort(timeReq, n, sizeof(int), cmp);
    long long left = timeReq[0], right = (long long)timeReq[0] + 20LL * splitTime;
    while (left < right) {
        long long mid = (left + right) / 2;
        if (check(timeReq, n, mid, splitTime)) right = mid;
        else left = mid + 1;
    }
    return left;
}
